design/nfs/nfs-controller-runtime.md

```diff
@@ -0,0 +1,157 @@
+# Implement controller-runtime in Rook NFS Operator
+
+## Background
+
+This proposal is to implement controller-runtime in Rook NFS Operator to improve reliability of the operator itself. Currently, Rook nfs-operator only simply watches an event of CustomResource from an informer using simple WatchCR method which has limited functionality such as event can not be re-queued if failed. To implement controller-runtime is expected to overcome the shortcomings of current implementation.
+
+## Why controller-runtime?
+
+Controller-runtime is widely used for writing Kubernetes operators. It is also leveraged by Kubebuilder and Operator SDK. Controller-runtime consists of several packages that have their respective responsibilities in building operators. The main function of controller-runtime is
+
+- **Manager:** Runnable for the operator with leader election option. It is also provides shared dependencies such as clients, caches, schemes, etc.
+- **Controller:** Provides types and functions for building Controllers which ensure for any given object, the actual state matches the desired state which called `Reconciling` process.
+- **Admission Webhook:** Provides methods to build an admission webhook (both Mutating Admission Webhook and Validating Admission Webhook) and bootstrap a webhook server.
+- **Envtest:** Provides libraries for integration testing by starting a local control plane (etcd and kube-apiserver).
+- **Matrics:** Provides metrics utility for controller.
+
+## Implementation
+
+The controller and webhook provided by controller-runtime is ultimately run by a `Manager`. So the first thing that needed is to replace `StartWatch` method with controller-runtime `Manager`. Once controller-runtime `Manager` configured, both controller and webhook can be registered. The easy way to construct a controller and webhook is using [Builder](https://godoc.org/sigs.k8s.io/controller-runtime#Builder).
+
```

Comment:
The CRD design and the deployment process is going to be same right?
If yes then add a line stating that.

Reply:
I just updated this section. please take a look. thanks!